### DISCUSSION TITLE
Future topics and features

### DISCUSSION THREAD

**Discussion by dsilhavy** [Dash-Industry-Forum/dash-live-source-simulator#108] [Ideas]
This thread can be used to discuss potential new features and topics as well as important changes and bugfixes. As a starter some ideas. We need to check what is already supported:

## Features
* Support for MPD Patching, see https://github.com/Dash-Industry-Forum/dash-live-source-simulator/pull/92
* Low Latency streaming
  * Resync element
  * PRFT support
* DRM
  *  ECCP support, requires Clearkey and a separate server endpoint
  *  Key rotation
  * Authz url
* DASH events
  * Version 0 vs Version 1 events
* `<ServiceDescription>`
  * `<OperatingBandwidth>`
  * All `<Latency>` parameters supported? (looks good)
  * All `<PlaybackRate>` parameters supported? (looks good)
* Dash Representations
  * `@qualityRanking`
* Negative test content
  * 404 errors
  * Corrupt segments
  * Gaps
  * negative/positive `@eptDelta`
  * negative/positive `@pdDelta`
  * Delayed requests
* `<Preroll>` element 
* Content steering

## Operational aspects

* Create a Docker image for easy setup and deployment of the live simulator
* Making a new livesim2 written in Go, see https://github.com/edgeware/mp4ff for potentially reusable code
* Upgrade to Python 3 needed?  // Python3 transition already done. The MPD Patch PR, is towards Python2 version, though.
* Content ingest is fragile, and limited in the sense that segment durations must be integer seconds in average. A more flexible approach that can support other segment durations would be helpful.

**Comment by wilaw** [comment]
Dynamic ad insertion via SSAI should also be a feature, as this is on the most difficult features to enable at low latency.

**Comment by irajs** [comment]
*  DASH events:
   * verification of correct MPD & inband event processing/dispatch mode/API
   * adding support for alternative MPD event (for support of preroll/midroll)

* onRequest XML processing:  
  *  verification of correct resolution (one-time resolution and condition for re-resolving with MPD update, etc.)

**Comment by dsilhavy** [comment]
Just saw this:
* https://github.com/Dash-Industry-Forum/test-vectors-drm-authz-token-provider
* https://github.com/Dash-Industry-Forum/dash-live-source-simulator-drm-data-generator
